priv/perf/apps/load_test/lib/ethereum/nonce_tracker.ex

```diff
@@ -0,0 +1,42 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule LoadTest.Ethereum.NonceTracker do
+  @moduledoc """
+  Nonce tracker for sending ethereum transactions
+  """
+
+  alias ExPlasma.Encoding
+
+  def init() do
+    :ets.new(:nonce_tracker, [:set, :public, :named_table])
+  end
+
+  def get_next_nonce(address) do
+    if Enum.empty?(:ets.lookup(:nonce_tracker, address)) do
+      current_nonce =
+        address
+        |> Encoding.to_hex()
+        |> Ethereumex.HttpClient.eth_get_transaction_count("pending")
+        |> elem(1)
+        |> Encoding.to_int()
+
+      # it might happen that this is called more than once, but
```

Comment:
Currently it's not. Only `Faucet` uses that. But I can imagine it will be used by some load tests that interact with Ethereum, doing exits, challanges, piggybacks etc.

Reply:
hm... then I'm having difficulties understanding what this `if do` does/provides? can you elaborate?

Reply:
When there is no nonce set for the address, we need to get it from eth client to set a starting value. Hence the first branch of if.
If starting nonce is already set then we just get it without setting a default value.

Alternative solution would be to call eth client each time to get the nonce and then call `:ets.update_counter`. I didn't check that, but I guess the way it's coded now it's more performant.

Reply:
but if there's no nonce, wouldn't `eth_get_transaction_count ` for an address return `0` always?

Reply:
It depends on the address. If we reuse sender's address it's not 0.

Reply:
thanks for your patience!